packages/field/src/useField.ts

```diff
@@ -5,64 +5,84 @@
  * found at http://www.apache.org/licenses/LICENSE-2.0.
  */
 
-import { useMemo } from 'react';
-import { useUIDSeed } from 'react-uid';
+import { useCallback, useMemo } from 'react';
+import { useId } from '@zendeskgarden/container-utilities';
+import { IUseFieldProps, IUseFieldReturnValue } from './types';
 
-import { IUseFieldPropGetters } from './types';
-
-export function useField(idPrefix?: string): IUseFieldPropGetters {
-  const seed = useUIDSeed();
-  const prefix = useMemo(() => idPrefix || seed(`field_${PACKAGE_VERSION}`), [idPrefix, seed]);
+export const useField = ({
+  idPrefix,
+  hasHint,
+  hasMessage
+}: IUseFieldProps): IUseFieldReturnValue => {
+  const prefix = useId(idPrefix);
   const inputId = `${prefix}--input`;
   const labelId = `${prefix}--label`;
   const hintId = `${prefix}--hint`;
   const messageId = `${prefix}--message`;
 
-  const getLabelProps = ({ id = labelId, htmlFor = inputId, ...other } = {}) => {
-    return {
+  const getLabelProps = useCallback<IUseFieldReturnValue['getLabelProps']>(
+    ({ id = labelId, htmlFor = inputId, ...other } = {}) => ({
+      'data-garden-container-id': 'containers.field.label',
+      'data-garden-container-version': PACKAGE_VERSION,
       id,
       htmlFor,
-      'data-garden-container-id': 'containers.field',
-      'data-garden-container-version': PACKAGE_VERSION,
       ...other
-    } as any;
-  };
+    }),
+    [labelId, inputId]
+  );
 
-  const getInputProps = (
-    { id = inputId, ...other } = {},
-    { isDescribed = false, hasMessage = false } = {}
-  ) => {
-    return {
+  const getHintProps = useCallback<IUseFieldReturnValue['getInputProps']>(
+    ({ id = hintId, ...other } = {}) => ({
+      'data-garden-container-id': 'containers.field.hint',
+      'data-garden-container-version': PACKAGE_VERSION,
       id,
-      'aria-labelledby': labelId,
-      'aria-describedby':
-        isDescribed || hasMessage
-          ? ([] as string[])
-              .concat(isDescribed ? hintId : [], hasMessage ? messageId : [])
-              .join(' ')
-          : null,
       ...other
-    } as any;
-  };
+    }),
+    [hintId]
+  );
 
-  const getHintProps = ({ id = hintId, ...other } = {}) => {
-    return {
-      id,
-      ...other
-    } as any;
-  };
+  const getInputProps = useCallback<IUseFieldReturnValue['getInputProps']>(
+    ({ id = inputId, ...other } = {}) => {
+      const describedBy = [];
+
+      if (hasHint) {
+        describedBy.push(hintId);
+      }
+
+      if (hasMessage) {
+        describedBy.push(messageId);
+      }
 
-  const getMessageProps = ({ id = messageId, ...other } = {}) => {
-    return {
+      return {
+        'data-garden-container-id': 'containers.field.input',
+        'data-garden-container-version': PACKAGE_VERSION,
+        id,
+        'aria-labelledby': labelId,
```

Comment:
I know this is pre-existing, but I think it might be redundant given the` for`/`id` relationship between label and input? If I'm blanking on something obvious, no worries.

Reply:
Few things here:

- it would be strongly inadvisable to change this now as the hook has been tested and in production use since 2019
- the belt+suspenders approach is recommended as it handles corner cases (i.e. hidden label or a label being rendered `as` a different element)
- perhaps most important: we're going to want this association when we knock out `getInputProps({ htmlFor = null })` in the case of a non-editable combobox